python/ray/data/impl/shuffle.py

```diff
@@ -13,137 +11,182 @@
+        shuffle_map_metadata = []
+        for i, refs in enumerate(shuffle_map_out):
+            shuffle_map_metadata.append(refs[0])
+            shuffle_map_out[i] = refs[1:]
+
+        # Eagerly delete the input block references in order to eagerly release
+        # the blocks' memory.
+        del input_blocks_list
+        if clear_input_blocks:
+            input_blocks.clear()
+        shuffle_map_metadata = map_bar.fetch_until_complete(shuffle_map_metadata)
+        map_bar.close()
+
+        reduce_bar = ProgressBar("Shuffle Reduce", position=0, total=output_num_blocks)
+        shuffle_reduce_out = [
+            shuffle_reduce.options(**reduce_ray_remote_args, num_returns=2,).remote(
+                *self._reduce_args,
+                *[shuffle_map_out[i][j] for i in range(input_num_blocks)],
+            )
+            for j in range(output_num_blocks)
+        ]
+        # Eagerly delete the map block references in order to eagerly release
+        # the blocks' memory.
+        del shuffle_map_out
+        new_blocks, new_metadata = zip(*shuffle_reduce_out)
+        reduce_bar.block_until_complete(list(new_blocks))
+        new_metadata = ray.get(list(new_metadata))
+        reduce_bar.close()
+
+        stats = {
+            "map": shuffle_map_metadata,
+            "reduce": new_metadata,
+        }
+
+        return BlockList(list(new_blocks), list(new_metadata)), stats
+
+
+class ShufflePartitionOp(ShuffleOp):
+    """
+    Operator used for `random_shuffle` and `repartition` transforms.
+    """
+
+    def __init__(
+        self,
+        block_udf=None,
+        random_shuffle: bool = False,
+        random_seed: Optional[int] = None,
+    ):
+        super().__init__(
+            map_args=[block_udf, random_shuffle, random_seed],
+            reduce_args=[random_shuffle, random_seed],
+        )
+
+    @staticmethod
+    def map(
+        idx: int,
+        block: Block,
+        output_num_blocks: int,
+        block_udf: Optional[Callable[[Block], Iterable[Block]]],
+        random_shuffle: bool,
+        random_seed: Optional[int],
+    ) -> List[Union[BlockMetadata, Block]]:
+        stats = BlockExecStats.builder()
+        if block_udf:
+            # TODO(ekl) note that this effectively disables block splitting.
+            blocks = list(block_udf(block))
+            if len(blocks) > 1:
+                builder = BlockAccessor.for_block(blocks[0]).builder()
+                for b in blocks:
+                    builder.add_block(b)
+                block = builder.build()
+            else:
+                block = blocks[0]
         block = BlockAccessor.for_block(block)
 
-    slice_sz = max(1, math.ceil(block.num_rows() / output_num_blocks))
-    slices = []
-    for i in range(output_num_blocks):
-        slices.append(block.slice(i * slice_sz, (i + 1) * slice_sz, copy=True))
-
-    # Randomize the distribution order of the blocks (this matters when
```

Comment:
@ericl I also removed this because I don't think it's necessary? I think you wrote this comment before so maybe you can confirm. I wasn't sure what it meant exactly.

Reply:
I think it matters when the block sizes are small. Otherwise certain reducers get less data. Can you leave it in just for safety in these cases?

Reply:
I see, I guess it could matter if there are 0 or 1 rows in each partition...I'll add it back.